### PR TITLE
Performance improvement to validateState()

### PR DESCRIPTION
**Pull request by BoneffFC**
Performance optimisation of state validation function

### CODE REVIEW

File: src/StateMachine/StateMachine/StateMachine.php [hunk 1]
```diff
@@ -717,7 +717,7 @@ private function bindTransitionsToStates()
      */
     private function validateState($state)
     {
-        if (isset($state) && !in_array($state, $this->states)) {
+        if (isset($state) && !isset($this->states[$state])) {
```
Comment (olschaefer) [inline src/StateMachine/StateMachine/StateMachine.php]: How big is the difference?

Comment (BoneffFC) [reply src/StateMachine/StateMachine/StateMachine.php]: Around 13 times - screenshots in https://jira.fundingcircle.com/browse/PC-50.

**Comment by BoneffFC** [comment]
Well - if we review logs from NewRelic the issue is no longer occuring in NL, and from the profiles this was all I found. If I am correct logs should be made public to be shared ? Will post a few here.

**Comment by BoneffFC** [comment]
Logs prior to SM change I tested locally:
https://blackfire.io/profiles/f727d514-6a94-47af-bfb0-29c8fd7e4baa/graph
https://blackfire.io/profiles/460c1249-8edc-4df1-bfd1-acb83d2e2849/graph
https://blackfire.io/profiles/19f9bdfe-99e8-4678-a500-72a7d7eb86f1/graph

**Comment by ahuminiuc** [comment]
From one of the black fire reports: 
screen_shot_2018-05-04_at_14_14_01

**Comment by BoneffFC** [comment]
Yes - but still if you take a closer look I think it is slow because of the call inside it to the State Machine.
Here is a profile after the change in the SM:
https://blackfire.io/profiles/9242987e-1c06-4448-aece-f8e61cff26b0/graph

File: src/StateMachine/StateMachine/StateMachine.php (hunk 1 above)
Comment (sgerrand) [reply src/StateMachine/StateMachine/StateMachine.php]: Are you able to assert this performance improvement using a benchmarking tool of some kind?

Comment (BoneffFC) [reply src/StateMachine/StateMachine/StateMachine.php]: @sgerrand Please refer to the above screenshots. I have set all the links from Blackfire profiles provided in this PR back to private as I am not sure if they should stay public.
If they need to be further reviewed please advice on a good way to share them privately.

**Pull request merged**